### PR TITLE
[Fix]fix trt nms jetson

### PR DESCRIPTION
**Pull request by grimoire**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily receiving feedbacks. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Add bounds to avoid large resource usage.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a dependency on downstream projects of a newer version, this PR should be tested with all supported versions of downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: csrc/mmdeploy/backend_ops/tensorrt/common_impl/nms/allClassNMS.cu [hunk 1]
```diff
@@ -66,17 +68,19 @@ __device__ float jaccardOverlap(const Bbox<T_BBOX> &bbox1, const Bbox<T_BBOX> &b
...
-                                   T_BBOX *bbox_data,  // bbox_data should be float to preserve
-                                                       // location information
-                                   T_SCORE *beforeNMS_scores, int *beforeNMS_index_array,
-                                   T_SCORE *afterNMS_scores, int *afterNMS_index_array,
-                                   bool flipXY = false) {
+__global__ void
+#if __CUDA_ARCH__ == 620 || __CUDA_ARCH__ == 530
+__launch_bounds__(512)
+#endif
+    allClassNMS_kernel(const int num, const int num_classes, const int num_preds_per_class,
```
Comment (lvhan028) [inline csrc/mmdeploy/backend_ops/tensorrt/common_impl/nms/allClassNMS.cu:75]: fault indent

File: csrc/mmdeploy/backend_ops/tensorrt/common_impl/nms/allClassNMS.cu [hunk 2]
```diff
@@ -196,29 +200,18 @@ pluginStatus_t allClassNMS_gpu(cudaStream_t stream, const int num, const int num
...
+  kernel[t_size - 1]<<<GS, BS, 0, stream>>>(
       num, num_classes, num_preds_per_class, top_k, nms_threshold, share_location, isNormalized,
       (T_BBOX *)bbox_data, (T_SCORE *)beforeNMS_scores, (int *)beforeNMS_index_array,
       (T_SCORE *)afterNMS_scores, (int *)afterNMS_index_array, flipXY);
 
   cudaError_t code = cudaGetLastError();
-  if (code != cudaSuccess) {
-    // Verify if cuda dev0 requires top_k to be reduced;
-    // sm_53 (Jetson Nano) and sm_62 (Jetson TX2) requires reduced top_k < 1000
-    auto __cuda_arch__ = get_cuda_arch(0);
```
Comment (lvhan028) [inline csrc/mmdeploy/backend_ops/tensorrt/common_impl/nms/allClassNMS.cu:213]: Should we consider `top_k < 1000` in jetson cases?

File: csrc/mmdeploy/backend_ops/tensorrt/common_impl/nms/allClassNMS.cu [hunk 3]
```diff
@@ -196,29 +200,18 @@ pluginStatus_t allClassNMS_gpu(cudaStream_t stream, const int num, const int num
...
-    if ((__cuda_arch__ == 530 || __cuda_arch__ == 620) && top_k >= 1000) {
-      printf(
-          "Warning: pre_top_k need to be reduced for devices with arch 5.3, 6.2, got "
-          "pre_top_k=%d\n",
-          top_k);
-    }
-  }
-  CSC(cudaGetLastError(), STATUS_FAILURE);
+  CUASSERT(code)
```
Comment (lvhan028) [inline csrc/mmdeploy/backend_ops/tensorrt/common_impl/nms/allClassNMS.cu:213]: better to put ";" at the end of this line

File: csrc/mmdeploy/backend_ops/tensorrt/common_impl/nms/allClassNMS.cu [hunk 4]
```diff
@@ -66,17 +68,19 @@ __device__ float jaccardOverlap(const Bbox<T_BBOX> &bbox1, const Bbox<T_BBOX> &b
...
-                                   const float nms_threshold, const bool share_location,
-                                   const bool isNormalized,
-                                   T_BBOX *bbox_data,  // bbox_data should be float to preserve
-                                                       // location information
-                                   T_SCORE *beforeNMS_scores, int *beforeNMS_index_array,
-                                   T_SCORE *afterNMS_scores, int *afterNMS_index_array,
-                                   bool flipXY = false) {
+__global__ void
+#if __CUDA_ARCH__ == 620 || __CUDA_ARCH__ == 530
+__launch_bounds__(512)
```
Comment (lvhan028) [inline csrc/mmdeploy/backend_ops/tensorrt/common_impl/nms/allClassNMS.cu:73]: What benefits can `__launch_bounds__` bring?

File: csrc/mmdeploy/backend_ops/tensorrt/common_impl/nms/allClassNMS.cu [hunk 5]
```diff
@@ -66,17 +68,19 @@ __device__ float jaccardOverlap(const Bbox<T_BBOX> &bbox1, const Bbox<T_BBOX> &b
...
-                                   const int num_preds_per_class, const int top_k,
-                                   const float nms_threshold, const bool share_location,
-                                   const bool isNormalized,
-                                   T_BBOX *bbox_data,  // bbox_data should be float to preserve
-                                                       // location information
-                                   T_SCORE *beforeNMS_scores, int *beforeNMS_index_array,
-                                   T_SCORE *afterNMS_scores, int *afterNMS_index_array,
-                                   bool flipXY = false) {
+__global__ void
+#if __CUDA_ARCH__ == 620 || __CUDA_ARCH__ == 530
```
Comment (AllentDan) [inline csrc/mmdeploy/backend_ops/tensorrt/common_impl/nms/allClassNMS.cu:72]: Is it possible that `__CUDA_ARCH__` is undefined?

File: csrc/mmdeploy/backend_ops/tensorrt/common_impl/nms/allClassNMS.cu (hunk 2 above)
Comment (AllentDan) [inline csrc/mmdeploy/backend_ops/tensorrt/common_impl/nms/allClassNMS.cu:213]: may also delete the definition of `get_cuda_arch`

File: csrc/mmdeploy/backend_ops/tensorrt/common_impl/nms/allClassNMS.cu (hunk 5 above)
Comment (grimoire) [reply csrc/mmdeploy/backend_ops/tensorrt/common_impl/nms/allClassNMS.cu:72]: NO as long as we compile cuda with nvcc. https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html#virtual-architecture-macros

File: csrc/mmdeploy/backend_ops/tensorrt/common_impl/nms/allClassNMS.cu (hunk 4 above)
Comment (grimoire) [reply csrc/mmdeploy/backend_ops/tensorrt/common_impl/nms/allClassNMS.cu:73]: A Jetson device has fewer resources (register/shared memory) than a desktop device. Which leads to error `too many resources required to launch` when TSIZE is large.
https://stackoverflow.com/questions/26201172/cuda-too-many-resources-requested-for-launch
Add [launch bounds](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#launch-bounds) would limit the register usage (for each multiprocessor).

File: csrc/mmdeploy/backend_ops/tensorrt/common_impl/nms/allClassNMS.cu (hunk 2 above)
Comment (grimoire) [reply csrc/mmdeploy/backend_ops/tensorrt/common_impl/nms/allClassNMS.cu:213]: Not necessary any more.

File: csrc/mmdeploy/backend_ops/tensorrt/common_impl/nms/allClassNMS.cu (hunk 5 above)
Comment (AllentDan) [reply csrc/mmdeploy/backend_ops/tensorrt/common_impl/nms/allClassNMS.cu:72]: I used the codes [here](https://forums.developer.nvidia.com/t/cuda-arch-undefined/26274#nvcc-gencodearchcompute_20codesm_20compute_20-testcu-o-test-1) and compiled it with `nvcc -gencode arch=compute_50,code=sm_50 test.cu -o test`. The log gave me this: *test.cu(11): error: identifier "__CUDA_ARCH__" is undefined*

File: csrc/mmdeploy/backend_ops/tensorrt/common_impl/nms/allClassNMS.cu (hunk 2 above)
Comment (grimoire) [reply csrc/mmdeploy/backend_ops/tensorrt/common_impl/nms/allClassNMS.cu:213]: Not in this PR. RotatedNMS use get_cuda_arch too. And I am not sure if launch bounds works on rotated detection model.

File: csrc/mmdeploy/backend_ops/tensorrt/common_impl/nms/allClassNMS.cu (hunk 5 above)
Comment (AllentDan) [reply csrc/mmdeploy/backend_ops/tensorrt/common_impl/nms/allClassNMS.cu:72]: > I used the codes [here](https://forums.developer.nvidia.com/t/cuda-arch-undefined/26274#nvcc-gencodearchcompute_20codesm_20compute_20-testcu-o-test-1) and compiled it with `nvcc -gencode arch=compute_50,code=sm_50 test.cu -o test`. The log gave me this: _test.cu(11): error: identifier "**CUDA_ARCH**" is undefined_

I found that the codes used `__CUDA_ARCH__` inside the func. If I placed it outside, there would be no error.

Comment (grimoire) [reply csrc/mmdeploy/backend_ops/tensorrt/common_impl/nms/allClassNMS.cu:72]: > This macro can be used in the implementation of GPU functions for determining the virtual architecture for which it is currently being compiled. The host code (the non-GPU code) must not depend on it.

Host code would be passed to gcc or clang, which does not has the arch macro. Only device code would be compiled by nvcc.
try compile the code below:
```
#ifdef __CUDA_ARCH__
#warning "hello"
#endif
```

File: csrc/mmdeploy/backend_ops/tensorrt/common_impl/nms/allClassNMS.cu (hunk 1 above)
Comment (grimoire) [reply csrc/mmdeploy/backend_ops/tensorrt/common_impl/nms/allClassNMS.cu:75]: It is formated by clang-format, any advice?

Comment (lvhan028) [reply csrc/mmdeploy/backend_ops/tensorrt/common_impl/nms/allClassNMS.cu:77]: use comment "// clang-format off" and "// clang-format on". Here is an example: https://github.com/open-mmlab/mmdeploy/blob/85320df2b46ed9a4752f68767b7deed4efc866d9/csrc/mmdeploy/apis/c/mmdeploy/model.cpp#L3

**Review by lvhan028 — APPROVED** [review]

**Review by AllentDan — APPROVED** [review]

**Pull request merged**